engine/pkg/externalresource/s3/index.go

```diff
@@ -0,0 +1,309 @@
+}
+
+type indexFile struct {
+	PersistedFileSet map[model.ResourceName]struct{} `json:"persisted_file_set"`
+}
+
+func newIndexFile() *indexFile {
+	return &indexFile{
+		PersistedFileSet: map[model.ResourceName]struct{}{},
+	}
+}
+
+func (f *indexFile) SetPersisted(ident internal.ResourceIdent) bool {
+	if _, ok := f.PersistedFileSet[ident.Name]; ok {
+		return false
+	}
+	f.PersistedFileSet[ident.Name] = struct{}{}
+	return true
+}
+
+func (f *indexFile) UnsetPersisted(ident internal.ResourceIdent) bool {
+	if _, ok := f.PersistedFileSet[ident.Name]; !ok {
+		return false
+	}
+	delete(f.PersistedFileSet, ident.Name)
+	return true
+}
+
+func (f *indexFile) DeserializeFrom(bytes []byte) error {
+	if err := json.Unmarshal(bytes, &f.PersistedFileSet); err != nil {
+		return errors.Annotate(err, "deserialize index file from s3")
+	}
+	return nil
+}
+
+func (f *indexFile) Serialize() ([]byte, error) {
+	bytes, err := json.Marshal(f.PersistedFileSet)
+	if err != nil {
+		return nil, errors.Annotate(err, "serialize index file to s3")
+	}
+	return bytes, nil
+}
+
+func (f *indexFile) GetPersistedFileSet() map[model.ResourceName]struct{} {
+	return f.PersistedFileSet
+}
+
+type indexManagerImpl struct {
+	executorID     model.ExecutorID
+	bucketSelector BucketSelector
+	storageFactory ExternalStorageFactory
+
+	indexMu     sync.Mutex
+	indexFiles  map[model.WorkerID]*indexFile
+	updateCount int
+
+	wg                 sync.WaitGroup
+	flushIndexNotifyCh chan struct {
+		project  tenant.ProjectInfo
+		workerID model.WorkerID
+		doneCh   chan error
+	}
+	closeCh  chan struct{}
+	isClosed atomic.Bool
+}
+
+func newIndexManager(
+	executorID model.ExecutorID,
+	bucketSelector BucketSelector,
+	factory ExternalStorageFactory,
+) *indexManagerImpl {
+	ret := &indexManagerImpl{
+		executorID:     executorID,
+		bucketSelector: bucketSelector,
+		storageFactory: factory,
+
+		indexFiles: make(map[model.WorkerID]*indexFile),
+
+		// a maximum of one pending notification is enough.
+		flushIndexNotifyCh: make(chan struct {
+			project  tenant.ProjectInfo
+			workerID model.WorkerID
+			doneCh   chan error
+		}, 1),
+		// closeCh is closed when we want the background task to exit.
+		closeCh: make(chan struct{}),
+	}
+
+	ret.wg.Add(1)
+	go func() {
+		defer ret.wg.Done()
+		ret.backgroundTask()
+	}()
+	return ret
+}
+
+func (m *indexManagerImpl) SetPersisted(ctx context.Context, ident internal.ResourceIdent) (bool, error) {
+	if m.isClosed.Load() {
+		return false, errors.New("indexManager already closed")
+	}
+
+	if m.executorID != ident.Executor {
+		log.Panic("Cannot persist resource that is not created on the local executor",
+			zap.Any("ident", ident))
+	}
+
+	m.indexMu.Lock()
+	index, ok := m.indexFiles[ident.WorkerID]
+	if !ok {
+		index = newIndexFile()
+		m.indexFiles[ident.WorkerID] = index
+	}
+
+	ok = index.SetPersisted(ident)
+	m.indexMu.Unlock()
+	if !ok {
+		return false, nil
+	}
+
+	if err := m.triggerFlush(ctx, ident.WorkerID); err != nil {
```

Comment:
Each persisted file will trigger one file flush to s3, do we need to introduce some batch mechanism for index file flush?